doc/poddoc/man/xtftp.8
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.13)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "XTFTP 8"
.TH XTFTP 8 "2009-11-16" "GNUspool Release 1" "GNUspool Print Manager"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
xtftp \- FTP protocol server driver
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBxtftp\fR
\&\fI\-h hostname\fR
[ \fI\-p port\fR ]
[ \fI\-d\fR ]
[ \fI\-u username\fR ]
[ \fI\-w password\fR ]
[ \fI\-o outfile\fR ]
[ \fI\-i infile\fR ]
[ \fI\-D directory\fR ]
[ \fI\-T timeout\fR ]
[ \fI\-A host\fR ]
[ \fI\-T\fR ]
[ \fI\-B\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBxtftp\fR is a terminal server program for use by \fBGNUspool\fR where
the destination is a server which accepts \s-1FTP\s0 protocol. Data is taken
from standard input, unless overridden by the \fB\-i\fR option.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\-h \fIhostname\fR" 8
.IX Item "-h hostname"
This is the host name or \s-1IP\s0 address to be used to access the
printer. In \fBGNUspool\fR setup files, it may be appropriate to
substitute \f(CW$SPOOLDEV\fR or \f(CW$SPOOLPTR\fR here.
.IP "\-A \fIhostname\fR" 8
.IX Item "-A hostname"
Specifies the sending hostname or \s-1IP\s0 address to be as given. This may
be required when sending from hosts with multiple \s-1IP\s0 addresses.
.IP "\-p \fIport number\fR" 8
.IX Item "-p port number"
This is the port number to be used, if not \fIftp\fR.
.IP "\-u \fIuser\fR" 8
.IX Item "-u user"
This is the user name to be quoted to the \s-1FTP\s0 server.
.IP "\-w \fIword\fR" 8
.IX Item "-w word"
This is the password to be quoted to the \s-1FTP\s0 server.
.IP "\-o \fIoutfile\fR" 8
.IX Item "-o outfile"
This is the output file to be stored at the server end. In some cases
this may designate the actual printer, and it may be useful to use
\&\f(CW$SPOOLDEV\fR as the argument to be expanded in \fBGNUspool\fR setup files.
.IP "\-D \fIdirectory\fR" 8
.IX Item "-D directory"
This is the output file to be stored in at the server end, if not the
\&\*(L"home directory\*(R". In some cases this may designate tha actual printer,
and it may be useful to use \f(CW$SPOOLDEV\fR as the argument to be
expanded in \fBGNUspool\fR setup files.
.IP "\-T \fItimeout\fR" 8
.IX Item "-T timeout"
This gives the timeout in milliseconds to await for responses from the
server.
.IP "\-i \fIinput\fR" 8
.IX Item "-i input"
Specify input file if not standard input.
.IP "\-t" 8
.IX Item "-t"
Specify text mode.
.IP "\-b" 8
.IX Item "-b"
Specify binary mode.
.IP "\-d" 8
.IX Item "-d"
Debug. Displays activities on standard error. Increasing information
is given if more than one \fB\-d\fR option is specified.
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
\&\fBxtftp\fR generates any appropriate diagnostics on standard error.
.SH "EXIT CODES"
.IX Header "EXIT CODES"
Normal termination is denoted by an exit code of zero.
.IP "1" 4
.IX Item "1"
Usage error, invalid option etc.
.IP "2" 4
.IX Item "2"
No host name or ip address given.
.IP "3" 4
.IX Item "3"
Cannot open input file.
.IP "4" 4
.IX Item "4"
Expecting password as well as user name.
.IP "5" 4
.IX Item "5"
Incorrect password.
.IP "6" 4
.IX Item "6"
Transmit failure \- file rejected.
.IP "18" 4
.IX Item "18"
Data accept failed.
.IP "41" 4
.IX Item "41"
Cannot create/bind control socket.
.IP "42" 4
.IX Item "42"
Cannot connect control socket.
.IP "44" 4
.IX Item "44"
Cannot create data socket.
.IP "45" 4
.IX Item "45"
Cannot bind data socket.
.IP "46" 4
.IX Item "46"
Cannot listen data socket.
.IP "47" 4
.IX Item "47"
Cannot get socket name data socket.
.IP "50" 4
.IX Item "50"
Unknown service name
.IP "51" 4
.IX Item "51"
Invalid \s-1IP\s0 address
.IP "52" 4
.IX Item "52"
Invalid host name
.IP "60" 4
.IX Item "60"
Socket write error.
.IP "61" 4
.IX Item "61"
Socket read error.
.IP "62" 4
.IX Item "62"
No response code from server.
.IP "71" 4
.IX Item "71"
Port or store command failed.
.IP "100" 4
.IX Item "100"
Generated command too long (excessive length file name etc).
.IP "200" 4
.IX Item "200"
Aborted with a signal.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIxilp\fR\|(8),
\&\fIxtelnet\fR\|(8),
\&\fIxtlpc\fR\|(8),
\&\fIxtlhp\fR\|(8).
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2009 Free Software Foundation, Inc.
This is free software. You may redistribute copies of it under the
terms of the \s-1GNU\s0 General Public License
<http://www.gnu.org/licenses/gpl.html>.
There is \s-1NO\s0 \s-1WARRANTY\s0, to the extent permitted by law.
.SH "AUTHOR"
.IX Header "AUTHOR"
John M Collins, Xi Software Ltd.


debian/changelog
gnuspool (1.7) unstable; urgency=low

  * strip out unneeded sections of postinst etc. and do some related tidying
  * remove MSWIN subdirectory and bump version in configure

 -- John M Collins (Xi Software Ltd) <jmc@xisl.com>  Mon, 26 Apr 2010 12:26:21 +0000

gnuspool (1.6) unstable; urgency=low

  [ John M Collins (Xi Software Ltd) ]
  * Initial release (Closes: #556996)

 -- John M Collins (Xi Software Ltd) <jmc@xisl.com>  Tue, 17 Nov 2009 10:16:21 +0000


src/termserv/xtftp.c
/* xtftp.c -- FTP printer driver

   Copyright 2008 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include "config.h"
#include <stdio.h>
#include <ctype.h>
#include <sys/types.h>
#ifdef	HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif
#ifdef	HAVE_SYS_FILIO_H
#include <sys/filio.h>
#endif
#ifdef	HAVE_FCNTL_H
#include <fcntl.h>
#endif
#ifdef	TIME_WITH_SYS_TIME
#include <sys/time.h>
#include <time.h>
#elif	defined(HAVE_SYS_TIME_H)
#include <sys/time.h>
#else
#include <time.h>
#endif
#include "defaults.h"
#include "incl_net.h"
#include "incl_unix.h"
#include "incl_sig.h"
#include "ecodes.h"
#include "files.h"

char	debug = 0;
char	*progname;
int	timeout = 750, main_timeout = 30000;
netid_t	myhostid;
SHORT	myport;
int	csock;

#define	IBUF_LNG	200

void	pushout(int sock, char *buf, int olen)
{
	while  (olen > 0)  {
		int	nb = write(sock, buf, olen);
		if  (nb < 0)  {
			fprintf(stderr, "%s: Socket write error\n", progname);
			exit(60);
		}
		buf += nb;
		olen -= nb;
	}
}

int	fill_buf(int sock, char *buf)
{
	int	ilng = 0;
	do  {
		int  il = read(sock, buf, 1);
		if  (il <= 0)  {
			fprintf(stderr, "%s: Socket read error\n", progname);
			exit(61);
		}
		ilng++;
	}  while  (*buf++ != '\n'  &&  ilng < IBUF_LNG);

	if  (ilng < IBUF_LNG)
		*buf = '\0';
	return  ilng;
}

void	getcode(int *code, char *buf)
{
	if  (isdigit(buf[0])  &&  isdigit(buf[1]) &&  isdigit(buf[2]) && isspace(buf[3]))
		*code = atoi(buf) / 100;
}

int	getresp(int sock, int ismulti)
{
	int	ilng, lastcode = 0;
	fd_set	ready;
	struct	timeval	tvs;
	char	ibuf[IBUF_LNG];

	FD_ZERO(&ready);
	FD_SET(sock, &ready);
	/* Treat timeout as milliseconds */
	tvs.tv_sec = main_timeout / 1000;
	tvs.tv_usec = (main_timeout % 1000) * 1000;

	if  (debug)
		fprintf(stderr, "Getting response...");

	if  (select(sock+1, &ready, (fd_set *) 0, (fd_set *) 0, &tvs) > 0)  {
		ilng = fill_buf(sock, ibuf);
		if  (debug)  {
			fputs("\n\t", stderr);
			fwrite(ibuf, 1, ilng, stderr);
		}
		getcode(&lastcode, ibuf);
	}
	else  {
		fprintf(stderr, "No response from server\n");
		exit(62);
	}

	if  (!ismulti)  {
		if  (debug  &&  lastcode != 0)
			fprintf(stderr, "Returning code %d\n", lastcode);
		return  lastcode;
	}

	for  (;;)  {

		tvs.tv_sec = timeout / 1000; /* Linux resets it */
		tvs.tv_usec = (timeout % 1000) * 1000;

		if  (select(sock+1, &ready, (fd_set *) 0, (fd_set *) 0, &tvs) <= 0)  {
			if  (debug  &&  lastcode != 0)
				fprintf(stderr, "Returning code %d\n", lastcode);
			return  lastcode;
		}

		ilng = fill_buf(sock, ibuf);

		if  (debug)  {
			putc('\t', stderr);
			fwrite(ibuf, 1, ilng, stderr);
		}

		getcode(&lastcode, ibuf);
	}
}

int  command(int sock, char *a1, char *a2, int multi)
{
	char	*bp;
	char	obuf[100];

	bp = obuf;
	strcpy(bp, a1);
	bp += strlen(a1);
	if  (a2)  {
		int	lng = strlen(a2);
		if  (bp - obuf + lng >= sizeof(obuf))  {
			fprintf(stderr, "%s: Arg %s too long\n", progname, a2);
			exit(100);
		}
		*bp++ = ' ';
		strcpy(bp, a2);
		bp += lng;
	}
	if  (debug)  {
		*bp = '\0';
		fprintf(stderr, "Sending command: %s\n", obuf);
	}

	*bp++ = '\r';
	*bp++ = '\n';
	pushout(sock, obuf, bp - obuf);
	if  (strcmp(a1, "QUIT") == 0)
		return  0;
	return  getresp(sock, multi);
}

RETSIGTYPE	catchabort(int n)
{
#ifdef	UNSAFE_SIGNALS
	signal(n, SIG_IGN);
#endif
	command(csock, "QUIT", (char *) 0, 0);
	exit(E_SIGNAL);
}

void	portcommand(int sock)
{
	int	cnt;
	char	*thing;
	unsigned  char	*cp;
	char	xbuf[40];

	cp = (unsigned char *) &myhostid;
	thing = xbuf;
	for  (cnt = 0;  cnt < sizeof(myhostid);  cnt++)
		thing += sprintf(thing, "%u,", *cp++);
	cp = (unsigned char *) &myport;
	sprintf(thing, "%u,%u", cp[0], cp[1]);
	if  (command(sock, "PORT", xbuf, 0) != 2)  {
		fprintf(stderr, "Port command failed\n");
		exit(71);
	}
}

SHORT	gserv(char *name)
{
	struct  servent	*sp;

	if  (isdigit(name[0]))
		return  htons((SHORT) atoi(name));

	if  (!(sp = getservbyname(name, "tcp"))  &&  !(sp = getservbyname(name, "TCP")))  {
		fprintf(stderr, "%s: Unknown service %s\n", progname, name);
		exit(50);
	}
	return  sp->s_port;
}

netid_t	ghost(char *name)
{
	if  (isdigit(name[0]))  {
		netid_t	hostid;
#ifdef	DGAVIION
		struct	in_addr  ina_str;
		ina_str = inet_addr(name);
		hostid = ina_str.s_addr;
#else
		hostid = inet_addr(name);
#endif
		if  (hostid == -1L)  {
			fprintf(stderr, "%s: Invalid internet address %s\n", progname, name);
			exit(51);
		}
		return  hostid;
	}
	else  {
		struct	hostent  *host;
		if  (!(host = gethostbyname(name)))  {
			fprintf(stderr, "%s: unknown host %s\n", progname, name);
			exit(52);
		}
		return  * (netid_t *) host->h_addr;
	}
}

void	getmyhostid(void)
{
	char	myname[256];

	myhostid = 0L;
	myname[sizeof(myname) - 1] = '\0';
	gethostname(myname, sizeof(myname) - 1);
	myhostid = ghost(myname);
}

void  initsockets(netid_t hostid, SHORT portnum, int *csock, int *dsock)
{
	int	cs, ds;
	SOCKLEN_T	sfrom;
	struct	sockaddr_in	csin, dsin, from;
#ifdef	SO_REUSEADDR
	int	on = 1;
#endif
	BLOCK_ZERO(&csin, sizeof(csin));
	BLOCK_ZERO(&dsin, sizeof(dsin));
	csin.sin_family = dsin.sin_family = AF_INET;
	csin.sin_addr.s_addr = hostid;
	dsin.sin_addr.s_addr = INADDR_ANY;
	csin.sin_port = portnum;
	dsin.sin_port = 0;

	if  ((cs = socket(AF_INET, SOCK_STREAM, 0)) < 0)  {
		fprintf(stderr, "%s: Cannot create control socket\n", progname);
		exit(41);
	}

	if  (connect(cs, (struct sockaddr *) &csin, sizeof(csin)) < 0)  {
		fprintf(stderr, "%s: Cannot connect control socket\n", progname);
		exit(42);
	}

	if  ((ds = socket(AF_INET, SOCK_STREAM, 0)) < 0)  {
		fprintf(stderr, "%s: Cannot create data socket\n", progname);
		exit(44);
	}
#ifdef	SO_REUSEADDR
	setsockopt(ds, SOL_SOCKET, SO_REUSEADDR, (char *) &on, sizeof(on));
#endif
	if  (bind(ds, (struct sockaddr *) &dsin, sizeof(dsin)) < 0)  {
		fprintf(stderr, "%s: Cannot bind data socket\n", progname);
		exit(45);
	}

	if  (listen(ds, 1) < 0)  {
		fprintf(stderr, "%s: Cannot listen on data socket\n", progname);
		exit(46);
	}

	sfrom = sizeof(from);
	if  (getsockname(ds, (struct sockaddr *) &from, &sfrom) < 0)  {
		fprintf(stderr, "%s: Cannot get sock name\n", progname);
		exit(47);
	}

	myport = from.sin_port;
	*csock = cs;
	*dsock = ds;
}

int	accept_data(int dsock)
{
	struct	sockaddr_in	from;
	SOCKLEN_T	fromlen = sizeof(from);
	return  accept(dsock, (struct sockaddr *) &from, &fromlen);
}

int	xmit_stdin(int dsock, char *outfile)
{
	int	cnt, ssock;
	char	xbuf[1024];

	if  ((cnt = fread(xbuf, 1, sizeof(xbuf), stdin)) <= 0)
		return  0;	/* Nothing to send */

	if  (command(csock, "STOR", outfile, 1) > 2)  {
		fprintf(stderr, "Store command failed\n");
		exit(71);
	}
	ssock = accept_data(dsock);

	if  (ssock < 0)  {
		fprintf(stderr, "%s: Cannot do accept\n", progname);
		exit(18);
	}

	do  {
		pushout(ssock, xbuf, cnt);
		cnt = fread(xbuf, 1, sizeof(xbuf), stdin);
	}  while  (cnt > 0);

	close(ssock);
	return  1;
}

MAINFN_TYPE	main(int argc, char **argv)
{
	char	*hostname = (char *) 0, *cport = (char *) 0;
	char	*username = (char *) 0, *password = (char *) 0;
	char	*outfile = (char *) 0, *infile = (char *) 0;
	char	*directory = (char *) 0;
	int	textmode = 0, c;
	int	dsock;
	extern	char	*optarg;
	char	outname[20];
#ifdef	STRUCT_SIG
	struct	sigstruct_name	zs;
#endif

	versionprint(argv, "$Revision: 1.1 $", 1);

	if  ((progname = strrchr(argv[0], '/')))
		progname++;
	else
		progname = argv[0];

	while  ((c = getopt(argc, argv, "h:H:p:P:du:w:o:i:tbA:T:R:D:")) != EOF)
		switch  (c)  {
		default:
			fprintf(stderr, "%s: invalid option\n", progname);
			return  1;
		case  'h':case  'H':
			hostname = optarg;
			continue;
		case  'p':case  'P':
			cport = optarg;
			continue;
		case  'd':
			debug++;
			continue;
		case  'u':
			username = optarg;
			continue;
		case  'w':
			password = optarg;
			continue;
		case  'o':
			outfile = optarg;
			continue;
		case  'i':
			infile = optarg;
			continue;
		case  't':
			textmode = 1;
			continue;
		case  'b':
			textmode = 0;
			continue;
		case  'D':
			directory = optarg;
			continue;
		case  'T':
			timeout = atoi(optarg);
			continue;
		case  'R':
			main_timeout = atoi(optarg);
			continue;
		case  'A':
			myhostid = ghost(optarg);
			continue;
		}

	if  (!hostname)  {
		fprintf(stderr, "%s: You must give a host name with -h\n", progname);
		return  2;
	}

	if  (myhostid  ==  0L)
		getmyhostid();

	if  (!cport)
		cport = "ftp";

	if  (!outfile)  {
		sprintf(outname, "XTFTP.%.10d", getpid());
		outfile = outname;
	}

	if  (infile  &&  !freopen(infile, "r", stdin))  {
		fprintf(stderr, "%s: Cannot open input file %s\n", progname, infile);
		return  3;
	}

	initsockets(ghost(hostname), gserv(cport), &csock, &dsock);

#ifdef	STRUCT_SIG
	zs.sighandler_el = catchabort;
	sigmask_clear(zs);
	zs.sigflags_el = SIGVEC_INTFLAG;
	sigact_routine(SIGTERM, &zs, (struct sigstruct_name *) 0);
#else
	signal(SIGTERM, catchabort);
#endif

	if  (getresp(csock, 1) != 2)  {
		fprintf(stderr, "Unexpected response from server\n");
		return  48;
	}

	if  (username)  {
		int  code = command(csock, "USER", username, 0);
		if  (code  ==  3)  {
			if  (!password)  {
				fprintf(stderr, "You must give a password for user %s\n", username);
				return  4;
			}
			code = command(csock, "PASS", password, 0);
			if  (code != 2)  {
				fprintf(stderr, "%s: Password not accepted\n", progname);
				return  5;
			}
		}
	}

	if  (directory  &&  command(csock, "CWD", directory, 0) != 2)  {
		fprintf(stderr, "%s: cannot select directory %s\n", progname, directory);
		return  7;
	}

	portcommand(csock);
	command(csock, "TYPE", textmode? "A": "I", 0);
	if  (xmit_stdin(dsock, outfile)  &&  getresp(csock, 0) != 2)  {
		fprintf(stderr, "Transmit failure\n");
		return  6;
	}
	command(csock, "QUIT", (char *) 0, 0);
	return  0;
}
